Use `analyseIfAddressTaken` logic for checking if address is taken in converter

ire4ever1190 opened the pull request:
Closes #21531 

This performs the analyses later on when a calls arguments are being checked. The two times when it wasn't properly being checked which lead to the `sfAddrTaken` not being set were
- During certain magic calls ([listed here](https://github.com/nim-lang/Nim/blob/172e2e1afc54c95b729f58aaf4d70802725a64c2/compiler/semexprs.nim#L793)) which ignored the rest of the checks
- When the `nkHiddenConv` node was hidden behind a `nkHiddenDeref`

ire4ever1190 committed:
Add a test case

There are way more test cases (See all branches of analyseIfAddressTaken but this covers at least a second branch

ire4ever1190 committed:
Port analyseIfAddressTaken from semexprs to sigmatch

This was done since we cannot import sem or semexprs (circular import) but we need the rest of the logic. In needs to be done here since the converter isn't semmed afterwards and so we can't just leave the process til later use the version from semexprs

Araq commented:
Sorry but this is getting too hacky. Remove the `sfAddrTaken` flag instead entirely and the backends that need it should compute it. Easier said than done, I know, but these hacks live on for good and end up causing more effort than a cleaner solution that initially takes more effort.

ire4ever1190 committed:
Less hacky solution which has the checking be done in analyseIfAddressTakenInCall

This was done instead of the recommendation on removing it since sfAddrTaken is used in places other than the backend

ire4ever1190 committed:
Remove weird whitespace

ire4ever1190 commented:
I looked into removing `sfAddrTaken` but it looks to be used in places other than the JS backend so don't know how possible that is.

I've cleaned up my fix (and updated my analysis) and believe it isn't too hacky anymore since it now uses the same method of checking that was used before but covers more edgecases (listed in analysis)

ire4ever1190 committed:
Still check nkHiddenAddr if we are checking a converter
